### ISSUE TITLE
Synthesizing tiles with an image generator

### ISSUE THREAD

**Issue by defgsus** [BorisTheBrave/resynth-tiles#3]
Hi @BorisTheBrave , 

first of all, thanks for all your [articles](https://www.boristhebrave.com/). 

I recently tried to use Diffusion models to create game tilesets. I tried this previously without great success and now i thought i do something unusual and, before experimenting myself, actually search the web if someone has done something similar.

Found your [article](https://www.boristhebrave.com/2025/02/04/generating-tilesets-with-stable-diffusion/) and agree on all the problems of controllability.  I even trained some diffusion models myself on a [tile dataset](https://github.com/defgsus/pixelart-dataset). They create nice tile graphics but in an uncontrollable way. So i went back to a more controllable algorithm, using CLIP as an image generator. It's working pretty well! Below is a link to a little write-up. Hope you enjoy.

Best regards!

https://defgsus.github.io/nn-experiments/html/posts/2025/clipig-tilesets.html

**Comment by BorisTheBrave** [comment]
Looks very cool.

Are you just doing a single generation with the entire tilemap as a mask?

It doesn't look like I wrote it down properly, but my old gimp plugin actually worked differently - it generates the tiles one at a time, in a fashion that ensures they tile.

**Comment by defgsus** [comment]
For the images in the article i usually did about 1,000 generation steps, with a small learnrate and batch size between 1 and 4. At each step (and for each batch image) i generate a different random map to show to the CLIP model. 

For example, using the 7x7 edge+corner tiling with a tilesize of 16x16 pixels, that gives a 112x112 pixels source image (filled with noise or nothing or some template graphics).

CLIP has an input size of 224² pixels, so i usually generate a random tiled map with 15x15 tiles (15*16 = 240) and randomly crop a 224² area from that to show it to CLIP and then backpropagate the error between current and target embedding (from the text or image prompt). And then repeat this a 1000 times. 

Since many tiles are used more than once in the randomly generated map, there is already some mini-batch processing happening in a single backpropagation step. And since tiles usually match to a couple of different neighbor tiles, over time, they all seamlessly melt together.

When doing the masking, with "inside" and "outside" map areas, i do the same thing as above, but reset the pixels in the source image that are not inside the mask after each backprop step. If you define some transitional area for the mask borders, then they'll get a mix of both target embeddings, which CLIP often handles quite "meaningfully."

I did understand your method in the diffusion tiles article, i guess. Yeah, there is unfortunately not much one can do to steer it in a certain direction. Unless maybe training a ControlNet for the exact purpose of generating seamless tiles based on mask templates.

From what i have seen, the diffusion generation fails even when just using a different noise at each diffusion step. I also tried to shift the image-to-diffuse some pixels horizontally and vertically (with wrapping around) at each diffusion step, to make them, at least, tileable with themselves, but without any luck.

The CLIP approach is very controllable, you can do whatever you want in between backprop steps, but the visual quality is still not were i would like to have it. It's also much slower compared to a small-res self-trained diffusion model. I mean, something like a less-than-10M parameter model. I guess one could skip the ControlNet part, if one trains the diffusion model in the first place. In any case, the training set is not the most easiest thing to create... 

And i could go on talking about these things for hours but there are other demands ;)
